File: backend/app/api/strategy_showcase.py
"""
戦略ショーケースAPI

自動生成された投資戦略のショーケース機能を提供するAPIエンドポイント
"""

from fastapi import APIRouter, HTTPException, Query
from typing import Dict, Any, List, Optional
from pydantic import BaseModel
import logging

from app.core.services.strategy_showcase_service import StrategyShowcaseService
from app.utils.api_response_utils import api_response, handle_api_exception

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/strategies", tags=["strategy-showcase"])

# グローバルサービスインスタンス
showcase_service = StrategyShowcaseService()


# リクエスト・レスポンスモデル
class StrategyListResponse(BaseModel):
    """戦略一覧レスポンス"""

    success: bool
    strategies: List[Dict[str, Any]]
    total_count: int
    message: str


class StrategyDetailResponse(BaseModel):
    """戦略詳細レスポンス"""

    success: bool
    strategy: Optional[Dict[str, Any]] = None
    message: str


class ShowcaseStatsResponse(BaseModel):
    """ショーケース統計レスポンス"""

    success: bool
    statistics: Dict[str, Any]
    message: str


# APIエンドポイント
@router.get("/showcase", response_model=StrategyListResponse)
async def get_showcase_strategies(
    category: Optional[str] = Query(None, description="戦略カテゴリでフィルタ"),
    risk_level: Optional[str] = Query(None, description="リスクレベルでフィルタ"),
    limit: Optional[int] = Query(None, ge=1, le=100, description="取得件数制限"),
    offset: int = Query(0, ge=0, description="オフセット"),
    sort_by: str = Query("expected_return", description="ソート項目"),
    sort_order: str = Query("desc", regex="^(asc|desc)$", description="ソート順序"),
):
    """
    ショーケース戦略一覧を取得

    フィルタリング、ソート、ページネーション機能付きで戦略一覧を取得します。
    """

    async def _get_strategies():
        strategies = showcase_service.get_showcase_strategies(
            category=category,
            risk_level=risk_level,
            limit=limit,
            offset=offset,
            sort_by=sort_by,
            sort_order=sort_order,
        )

        return api_response(
            data=strategies,
            message="戦略一覧を取得しました",
            additional_fields={"total_count": len(strategies)},
        )

    return await handle_api_exception(_get_strategies)


@router.get("/showcase/stats", response_model=ShowcaseStatsResponse)
async def get_showcase_statistics():
    """
    ショーケース統計情報を取得

    全戦略の統計情報（平均リターン、カテゴリ分布等）を取得します。
    """

    async def _get_stats():
        statistics = showcase_service.get_showcase_statistics()

        return api_response(data=statistics, message="統計情報を取得しました")

    return await handle_api_exception(_get_stats)


@router.get("/showcase/{strategy_id}", response_model=StrategyDetailResponse)
async def get_strategy_detail(strategy_id: int):
    """
    戦略詳細を取得

    指定されたIDの戦略の詳細情報を取得します。
    """

    async def _get_detail():
        strategy = showcase_service.get_strategy_by_id(strategy_id)

        if not strategy:
            raise HTTPException(status_code=404, detail="Strategy not found")

        return api_response(data=strategy, message="戦略詳細を取得しました")

    return await handle_api_exception(_get_detail)


@router.get("/categories")
async def get_strategy_categories():
    """
    利用可能な戦略カテゴリ一覧を取得
    """

    async def _get_categories():
        categories = {
            "trend_following": "トレンドフォロー",
            "mean_reversion": "逆張り",
            "breakout": "ブレイクアウト",
            "range_trading": "レンジ取引",
            "momentum": "モメンタム",
        }

        return api_response(
            data=categories,
            message="戦略カテゴリ一覧を取得しました",
        )

    return await handle_api_exception(_get_categories)


@router.get("/risk-levels")
async def get_risk_levels():
    """
    利用可能なリスクレベル一覧を取得
    """

    async def _get_risk_levels():
        risk_levels = {"low": "低リスク", "medium": "中リスク", "high": "高リスク"}

        return api_response(
            data=risk_levels,
            message="リスクレベル一覧を取得しました",
        )

    return await handle_api_exception(_get_risk_levels)

File: frontend/types/strategy-showcase.ts
/**
 * 戦略ショーケース関連の型定義
 *
 */

/**
 * 戦略カテゴリ
 */
export type StrategyCategory =
  | "trend_following"
  | "mean_reversion"
  | "breakout"
  | "range_trading"
  | "momentum";

/**
 * リスクレベル
 */
export type RiskLevel = "low" | "medium" | "high";

/**
 * ソート項目
 */
export type SortBy =
  | "expected_return"
  | "sharpe_ratio"
  | "max_drawdown"
  | "win_rate"
  | "created_at";

/**
 * ソート順序
 */
export type SortOrder = "asc" | "desc";

/**
 * ショーケース戦略データ
 */
export interface StrategyShowcase {
  id: number;
  name: string;
  description: string;
  category: StrategyCategory;
  indicators: string[];
  parameters: Record<string, any>;
  expected_return: number;
  sharpe_ratio: number;
  max_drawdown: number;
  win_rate: number;
  gene_data: Record<string, any>;
  backtest_result_id?: number;
  risk_level: RiskLevel;
  recommended_timeframe: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

/**
 * 戦略フィルター
 */
export interface StrategyFilter {
  category?: StrategyCategory;
  risk_level?: RiskLevel;
  min_return?: number;
  max_return?: number;
  min_sharpe?: number;
  max_sharpe?: number;
  min_drawdown?: number;
  max_drawdown?: number;
  search_query?: string;
}

/**
 * ページネーション
 */
export interface Pagination {
  limit?: number;
  offset: number;
  sort_by: SortBy;
  sort_order: SortOrder;
}

/**
 * 戦略一覧レスポンス
 */
export interface StrategyListResponse {
  success: boolean;
  strategies: StrategyShowcase[];
  total_count: number;
  message: string;
}

/**
 * 戦略詳細レスポンス
 */
export interface StrategyDetailResponse {
  success: boolean;
  strategy?: StrategyShowcase;
  message: string;
}

/**
 * ショーケース統計
 */
export interface ShowcaseStatistics {
  total_strategies: number;
  avg_return: number;
  avg_sharpe_ratio: number;
  avg_max_drawdown: number;
  category_distribution: Record<StrategyCategory, number>;
  risk_distribution: Record<RiskLevel, number>;
}

/**
 * 統計レスポンス
 */
export interface ShowcaseStatsResponse {
  success: boolean;
  statistics: ShowcaseStatistics;
  message: string;
}

/**
 * 戦略生成リクエスト
 */
export interface GenerateShowcaseRequest {
  count: number;
  base_config?: {
    symbol: string;
    timeframe: string;
    start_date: string;
    end_date: string;
    initial_capital: number;
    commission_rate: number;
  };
}

/**
 * 戦略生成レスポンス
 */
export interface GenerateShowcaseResponse {
  success: boolean;
  message: string;
  generated_count: number;
  saved_count: number;
}

/**
 * カテゴリ一覧レスポンス
 */
export interface CategoriesResponse {
  success: boolean;
  categories: Record<StrategyCategory, string>;
  message: string;
}

/**
 * リスクレベル一覧レスポンス
 */
export interface RiskLevelsResponse {
  success: boolean;
  risk_levels: Record<RiskLevel, string>;
  message: string;
}

/**
 * 戦略カードプロパティ
 */
export interface StrategyCardProps {
  strategy: StrategyShowcase;
  onViewDetail: (strategy: StrategyShowcase) => void;
  className?: string;
}

/**
 * 戦略フィルタープロパティ
 */
export interface StrategyFiltersProps {
  filter: StrategyFilter;
  pagination: Pagination;
  onFilterChange: (filter: StrategyFilter) => void;
  onPaginationChange: (pagination: Pagination) => void;
  categories: Record<StrategyCategory, string>;
  riskLevels: Record<RiskLevel, string>;
}

/**
 * 戦略モーダルプロパティ
 */
export interface StrategyModalProps {
  strategy?: StrategyShowcase;
  isOpen: boolean;
  onClose: () => void;
}

/**
 * パフォーマンス指標の色分け
 */
export interface PerformanceColors {
  return: "positive" | "negative" | "neutral";
  sharpe: "excellent" | "good" | "fair" | "poor";
  drawdown: "low" | "medium" | "high";
}

/**
 * 戦略統計サマリー
 */
export interface StrategySummary {
  best_return: StrategyShowcase;
  best_sharpe: StrategyShowcase;
  lowest_drawdown: StrategyShowcase;
  highest_win_rate: StrategyShowcase;
}

/**
 * API エラーレスポンス
 */
export interface ApiErrorResponse {
  success: false;
  message: string;
  detail?: string;
}

/**
 * ローディング状態
 */
export interface LoadingState {
  strategies: boolean;
  statistics: boolean;
  generation: boolean;
  detail: boolean;
}

/**
 * エラー状態
 */
export interface ErrorState {
  strategies?: string;
  statistics?: string;
  generation?: string;
  detail?: string;
}
